### PR TITLE
Realize concurrent repeated request verification based on sliding window mechanism, thus supporting RPC retry

### LINKED ISSUE #4177: Support load RPC retry
## Enhancement
StarRocks currently supports concurrent RPC data writing, but lacks verification of repeated data requests, so there may be problems with retrying.
Therefore, the concurrent repeated request verification is realized through the sliding window mechanism, thereby supporting RPC retry.

### PR DESCRIPTION
**Pull request by meegoo**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4177 

## Problem Summary(Required) ：

StarRocks currently supports concurrent RPC data writing, but lacks verification of repeated data requests, so there may be problems with retrying.
Therefore, the concurrent repeated request verification is realized through the sliding window mechanism, thereby supporting RPC retry.

### CODE REVIEW

File: be/src/runtime/tablets_channel.cpp [hunk 1]
```diff
@@ -104,6 +104,46 @@ void TabletsChannel::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChu
         return;
     }
 
+    {
+        std::lock_guard<bthread::Mutex> lock(_senders[request.sender_id()].lock);
```
Comment (sduzh) [inline be/src/runtime/tablets_channel.cpp:108]: Nit: 
```suggestion
        std::lock_guard lock(_senders[request.sender_id()].lock);
```

File: be/src/runtime/tablets_channel.h [hunk 2]
```diff
@@ -106,7 +106,12 @@ class TabletsChannel : public RefCountedThreadSafe<TabletsChannel> {
     static std::atomic<uint64_t> _s_tablet_writer_count;
 
     struct Sender {
-        std::mutex lock;
+        bthread::Mutex lock;
+
+        std::set<int64_t> _receive_sliding_window;
+        std::set<int64_t> _success_sliding_window;
+
+        int64_t _last_sliding_packet_seq = -1;
```
Comment (sduzh) [inline be/src/runtime/tablets_channel.h:114]:
```suggestion
        std::set<int64_t> receive_sliding_window;
        std::set<int64_t> success_sliding_window;

        int64_t last_sliding_packet_seq = -1;
```

File: be/src/runtime/tablets_channel.h [hunk 3]
```diff
@@ -199,6 +204,7 @@ class TabletsChannel : public RefCountedThreadSafe<TabletsChannel> {
     // next sequence we expect
     std::atomic<int> _num_remaining_senders;
     std::vector<Sender> _senders;
+    size_t _max_sliding_window_size = config::max_load_dop * 3;
```
Comment (sduzh) [inline be/src/runtime/tablets_channel.h:207]: Initialize this field variable in the initialization list, as other fields. Keep code style consistent within the same scope.

File: be/src/runtime/tablets_channel.cpp [hunk 4]
```diff
@@ -104,6 +104,46 @@ void TabletsChannel::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChu
         return;
     }
 
+    {
```
Comment (sduzh) [inline be/src/runtime/tablets_channel.cpp:107]: Nit: move this code fragment into a new method.

File: be/src/runtime/tablets_channel.cpp [hunk 5]
```diff
@@ -178,6 +218,22 @@ void TabletsChannel::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChu
     // This will only block the bthread, will not block the pthread
     count_down_latch.wait();
 
+    {
```
Comment (sduzh) [inline be/src/runtime/tablets_channel.cpp:221]: Nit: move this code fragment into a separated method.

File: be/src/runtime/tablets_channel.h (hunk 2 above)
Comment (sduzh) [reply be/src/runtime/tablets_channel.h:114]: According to our coding guide, the name of public variables should not start with an underscore.

**Review by sduzh — DISMISSED** [review]

**Review by chaoyli — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Pull request merged**